csrc/preseg_passes/reorder_sharded_axis.cpp

```diff
@@ -10,142 +10,178 @@
 #include <device_lower/utils.h>
 #include <fusion.h>
 #include <host_ir/lower.h>
+#include <ir/allocation_utils.h>
 #include <ir/base_nodes.h>
 #include <ir/interface_nodes.h>
 #include <ir/utils.h>
 #include <multidevice/utils.h>
 #include <ops/alias.h>
 #include <ops/arith.h>
+#include <scheduler/utils.h>
+#include <transform_replay.h>
 
 namespace nvfuser::preseg_passes {
 
+namespace {
+
+void transform_and_parallelize_like(TensorView* ref, TensorView* tv) {
+  auto old2new = reorderDIDToFront(ref);
+  propagateDIDTransform(
+      ref, {tv}, /*did_pos=*/old2new.size(), PropagateDirection::kForward);
+  shardAllLike(ref, {tv}, {kParallelTypeDIDs.begin(), kParallelTypeDIDs.end()});
+}
+
+void makeCommunicationLayoutCompliant(
+    Expr* expr,
+    CommunicationInfo communication_info) {
+  TensorView* input = expr->inputs().at(0)->as<TensorView>();
+  TensorView* output = expr->outputs().at(0)->as<TensorView>();
+
+  IterDomain* p_sharded_id = communication_info.p_sharded_id;
+  IterDomain* c_sharded_id = communication_info.c_sharded_id;
+
+  // Copy input if:
+  // 1. Input is not contiguous.
+  // 2. Input is not allocation compliant and is not a reduce/allreduce.
+  if (isTvContiguous(input) &&
+      (communication_info.type == CommunicationType::Reduce ||
+       isAllocationCompliant(input, p_sharded_id))) {
```

Comment:
Should isTvContiguous be part of isAllocationCompliant? Conceptually, contiguity is part of allocation so I'd do that to avoid confusion. That implies some logic changes in the caller, e.g., here `(is reduce and contiguous) or isAllocationCompliant`. 

Alternatively, we could name it isAllocationOrderCompliant. This way, it's understandable that it doesn't check contiguity.